grabbit/src/test/groovy/com/twcable/grabbit/jcr/ProtoPropertyDecoratorSpec.groovy

```diff
@@ -0,0 +1,210 @@
+package com.twcable.grabbit.jcr
+
+import com.google.protobuf.ByteString
+import com.twcable.grabbit.proto.NodeProtos.Property as PropertyProto
+import com.twcable.grabbit.proto.NodeProtos.Value as ProtoValue
+import com.twcable.grabbit.proto.NodeProtos.Values as ProtoValues
+import org.apache.jackrabbit.value.BinaryValue
+import org.apache.jackrabbit.value.DateValue
+import org.apache.jackrabbit.value.StringValue
+import spock.lang.Specification
+
+import javax.jcr.Node
+import javax.jcr.Property
+import javax.jcr.PropertyType
+import javax.jcr.Value
+import javax.jcr.ValueFormatException
+
+import static org.apache.jackrabbit.JcrConstants.JCR_MIXINTYPES
+import static org.apache.jackrabbit.JcrConstants.JCR_PRIMARYTYPE
+
+class ProtoPropertyDecoratorSpec extends Specification {
+
+    def "When a property is either jcr:primaryType or jcr:mixinTypes, decorator will refuse to write these nodes are regular properties"() {
+        given:
+        final mockNode = Mock(Node)
+
+        when:
+        PropertyProto someProperty = PropertyProto.newBuilder()
+                                        .setName(refusedType)
+                                        .setType(PropertyType.STRING)
+                                        .setValue(
+                                            ProtoValue.newBuilder()
+                                                .setStringValue("somevalue")
+                                                .build()
+                                        )
+                                        .build()
+
+        final propertyDecorator = new ProtoPropertyDecorator(someProperty)
+        propertyDecorator.writeToNode(mockNode)
+
+        then:
+        thrown(IllegalStateException)
+
+        where:
+        refusedType << [JCR_PRIMARYTYPE, JCR_MIXINTYPES]
+    }
+
+
+    def "A single-value property can be written correctly to a node that expects a single-value property"() {
+        given:
+        final mockNode = Mock(Node)
+
+        when:
+        PropertyProto someProperty = PropertyProto.newBuilder()
+                                            .setName("someproperty")
+                                            .setType(PropertyType.STRING)
+                                            .setValue(
+                                                ProtoValue.newBuilder()
+                                                    .setStringValue("somevalue")
+                                                    .build()
+                                            )
+                                            .build()
+
+        final propertyDecorator = new ProtoPropertyDecorator(someProperty)
+        propertyDecorator.writeToNode(mockNode)
+
+        then:
+        1 * mockNode.setProperty("someproperty", new StringValue("somevalue"), PropertyType.STRING)
```

Comment:
Just to avoid hardcoded values at multiple places:
```
when:
        PropertyProto someProperty = PropertyProto.newBuilder()
                                            .setName(property)
                                            .setType(PropertyType.STRING)
                                            .setValue(
                                                ProtoValue.newBuilder()
                                                    .setStringValue(value)
                                                    .build()
                                            )
                                            .build()

        final propertyDecorator = new ProtoPropertyDecorator(someProperty)
        propertyDecorator.writeToNode(mockNode)

        then:
        1 * mockNode.setProperty(property, new StringValue(value), PropertyType.STRING)

        where:
        property        | value
        "someproperty"  | "somevalue"
```